✨ Support curating specific Source versions

Zethson opened the pull request:
WIP

Fixes https://github.com/laminlabs/lamindb/issues/2864

- Now support curating against specific Source versions

```python
!lamin init --storage run-tests --modules bionty

import lamindb as ln
import bionty as bt
import pandas as pd

ln.track()

bt.Disease.add_source(bt.Source.using("laminlabs/bionty-assets").get(entity=f"bionty.Disease", version="2024-08-06", organism="all").save())

disease_ontology_old = bt.Source.get(
    entity="bionty.Disease", name="mondo", version="2024-08-06"
)

schema = ln.Schema(
    name="My EHR schema",
    features=[
        ln.Feature(
            name="disease",
            dtype=bt.Disease,
            default_value="normal",
            nullable=False,
            cat_filters={"source__uid": disease_ontology_old.uid},
        ).save(),
    ],
).save()

dataset = {
    "disease": pd.Categorical(
        [
            "HDAC4-related haploinsufficiency syndrome",  # new in the latest 2025 mondo
            "SAMD9L-related spectrum and myeloid neoplasm risk",  # new in the latest 2025 mondo
            "essential hypertension",
            "essential hypertension",
            "asthma",
        ]
    ),
}
df = pd.DataFrame(dataset)

curator = ln.curators.DataFrameCurator(df, schema)
try:
    curator.validate()
except ln.errors.ValidationError as e:
    print(e)

________________
! 2 terms not validated in feature 'disease': 'HDAC4-related haploinsufficiency syndrome', 'SAMD9L-related spectrum and myeloid neoplasm risk'
    → fix typos, remove non-existent values, or save terms via: curator.cat.add_new_from('disease')
```

sunnyosun commented:
You don't run CI locally, do you? The curator tests are super easy to run locally so you don't need to wait for CI to error:
<img width="848" alt="Screenshot 2025-06-27 at 17 44 08" src="https://github.com/user-attachments/assets/4ca9ce56-3477-4537-9783-3217a55a3ab2" />

falexwolf commented:
Essentially an adaptation of these tests to the general filters you allow: https://github.com/laminlabs/lamindb/blob/b62e9d22422d29e27d7a5b1b6b0ed837510e96ad/tests/core/test_dtype.py#L28-L58

Zethson commented:
@falexwolf Made the following changes:
1. Disentangled the parsing of the input str into a format that can be used for filtering and the filtering itself into two separate functions. These can be more easily tested and can be reused more broadly.
2. I moved these now two functions from `core` of `curators` into `feature`.
3. I generally made these functions more robust and added lots of additional tests.
4. `resolve_relation_filters` could generally be more explicit about error handling to bubble up issues but I'm fine with keeping it as is.

Let me know what you think! I'd like to solve outstanding issues (if any) immediately so that I can continue with CxG.

falexwolf commented:
These changes all look great!

I just don't understand why we still don't have the equivalent test to this:

https://github.com/laminlabs/lamindb/blob/b62e9d22422d29e27d7a5b1b6b0ed837510e96ad/tests/core/test_dtype.py#L28-L58

And correspondingly, a test that checks how the `cat_filters` argument is serialized as a string after a feature is created.

Don't we need that?

falexwolf commented on lamindb/models/feature.py:
```diff
@@ -460,35 +531,35 @@ class Feature(SQLRecord, CanCurate, TracksRun, TracksUpdates):
             Feature sets.
 
-    Example:
+    Example::
```
This will break the section formatting. Please revert.

falexwolf commented on lamindb/models/feature.py:
```diff
@@ -460,35 +531,35 @@ class Feature(SQLRecord, CanCurate, TracksRun, TracksUpdates):
-    Example:
+    Example::
 
         A simple `"str"` feature.
```
If you remove the `>>>`, you need to add a double colon after a "A simple `"str"` feature.

falexwolf reviewed (approved):
Now things are easy to understand (at least for me 😄).

falexwolf commented on tests/core/test_dtype.py:
```diff
@@ -207,6 +207,22 @@ def test_nested_cat_dtypes():
+    # Generated via:
+    # dtype_str = ln.Feature(name="disease", dtype=bt.Disease, cat_filters={"source": disease_ontology_old}).save().dtype
+    # which is found in test_feature.py 'test_cat_filters_dtype'
+    dtype_str = "cat[bionty.Disease[source__uid='4a3ejKuf']]"
```
Double-check: concatenation is through `,` between "expressions", right?
